packages/realm-server/billing/stripe-webhook-handlers/payment-succeeded.ts

```diff
@@ -0,0 +1,148 @@
+    // it is not guaranteed that the user will have a stripe customer id until we get to here. That's why we want to retry
+    // a couple of times with a delay in between.
+    let user = await retry(
+      () => getUserByStripeId(dbAdapter, event.data.object.customer),
+      { retries: 5, delayMs: 1000 },
+    );
+
+    if (!user) {
+      throw new Error(
+        `Unrecognized stripe customer id ${event.data.object.customer}`,
+      );
+    }
+
+    let billingReason: 'subscription_create' | 'subscription_cycle' =
+      event.data.object.billing_reason;
+
+    if (billingReason === 'subscription_create') {
+      let subscription = await insertSubscription(dbAdapter, {
+        user_id: user.id,
+        plan_id: plan.id,
+        started_at: event.data.object.period_start,
+        status: 'active',
+        stripe_subscription_id: event.data.object.subscription,
+      });
+
+      let subscriptionCycle = await insertSubscriptionCycle(dbAdapter, {
+        subscriptionId: subscription.id,
+        periodStart: event.data.object.period_start,
+        periodEnd: event.data.object.period_end,
+      });
+
+      await addToCreditsLedger(
+        dbAdapter,
+        user.id,
+        plan.creditsIncluded,
+        'plan_allowance',
+        subscriptionCycle.id,
+      );
+    } else if (billingReason === 'subscription_cycle') {
+      let currentActiveSubscription = await getCurrentActiveSubscription(
+        dbAdapter,
+        user.id,
+      );
+
+      if (!currentActiveSubscription) {
+        throw new Error(
+          'Should never get here: no active subscription found when renewing',
+        );
+      }
+
+      let currentSubscriptionCycle = await getMostRecentSubscriptionCycle(
+        dbAdapter,
+        currentActiveSubscription.id,
+      );
+
+      if (!currentSubscriptionCycle) {
+        throw new Error(
+          'Should never get here: no current subscription cycle found when renewing',
+        );
+      }
+
+      let creditsToExpire = await sumUpCreditsLedger(dbAdapter, {
+        creditType: ['plan_allowance', 'plan_allowance_used'],
+        subscriptionCycleId: currentSubscriptionCycle.id,
+      });
+
+      await addToCreditsLedger(
+        dbAdapter,
+        user.id,
+        -creditsToExpire,
+        'plan_allowance_expired',
+        currentSubscriptionCycle.id,
+      );
+
+      let newSubscriptionCycle = await insertSubscriptionCycle(dbAdapter, {
+        subscriptionId: currentActiveSubscription.id,
+        periodStart: event.data.object.period_start,
+        periodEnd: event.data.object.period_end,
+      });
+
+      await addToCreditsLedger(
+        dbAdapter,
+        user.id,
+        plan.creditsIncluded,
+        'plan_allowance',
+        newSubscriptionCycle.id,
+      );
+    }
+    await query(dbAdapter, [
+      `UPDATE stripe_events SET is_processed = TRUE WHERE stripe_event_id = `,
+      param(event.id),
+    ]);
```

Comment:
This is a long method. It is pretty clear, which is great, but might benefit from being broken down into some smaller methods with expressive names.